File: javascript-sdk/packages/core/src/services/socket_service.ts
import { io, Socket as SocketIO } from "socket.io-client";
import { PaymentOptions } from "../types/payment-client-types";

interface _SocketResponse<T = any> {
  success: boolean,
  event: string,
  message: string,
  result: T
}

export class SocketService {
  static wsUrl = "https://pb-test-ws.apolopay.app"

  private options: PaymentOptions;
  private socket: SocketIO | null = null;

  constructor(options: PaymentOptions) {
    this.options = options;
  }

  public connect(processId: string): void {
    if (typeof window === 'undefined') return;

    if (this.socket && this.socket.connected) return this.disconnect()

    this.socket = io(SocketService.wsUrl, {
      extraHeaders: {
        "x-public-key": this.options.publicKey
      },
      transports: ['polling']
    });

    this.socket.on('connect', () => this.socket?.emit('process:connect', { processId }));

    console.log(`Conectado a Socket.io para processId: ${processId}`);

    this.socket.on('process:message', (response: _SocketResponse) => this.handleWebSocketMessage(response));

    this.socket.on('connect_error', (error) => {
      console.error('Error en conexión Socket.io:', error);
      this.options.onError({ code: 'connect_error', message: 'Error de conexión en tiempo real.', error });
      this.disconnect();
    });

    this.socket.on('disconnect', (reason) => {
      console.info(`Socket.io Desconectado: ${reason}`);
      this.socket = null;
    });
  }

  private handleWebSocketMessage(response: _SocketResponse): void {
    console.log(response);

    if (!response.success) {
      return this.options.onError({
        code: 'payment_failed',
        message: response.message,
        error: response.result
      });
    }

    if (response.result.status === 'completed') {
      return this.options.onSuccess({
        code: 'payment_success',
        message: response.message,
        result: response.result
      });
    }
  }

  public disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
  }
}

File: dart_sdk/lib/services/socket_service.dart
import 'package:apolopay_sdk/apolopay_sdk.dart';
import 'package:apolopay_sdk/utils/variables.dart';
import 'package:flutter/material.dart';
import 'package:socket_io_client/socket_io_client.dart' as io;
import '../models/client_response.dart';

enum SocketEvents {
  fundsReceived('funds_received'),
  partialPayment('partial_payment');

  const SocketEvents(this.value);
  final String value;
}

class _SocketResponse<T> {
  final bool success;
  final SocketEvents event;
  final String message;
  final T result;

  _SocketResponse({
    required this.success,
    required this.event,
    required this.message,
    required this.result,
  });

  Map<String, dynamic> toJson() => {
        'success': success,
        'event': event,
        'message': message,
        'result': result,
      };

  factory _SocketResponse.fromJson(Map<String, dynamic> json) {
    return _SocketResponse(
      success: json['success'],
      event: json['event'],
      message: json['message'],
      result: json['result'],
    );
  }
}

class SocketService {
  SocketService(this.options);
  final ApoloPayOptions options;
  io.Socket? _socket;

  void connect(String processId) {
    if (_socket != null && _socket!.connected) return;

    _socket = io.io(
      socketURL,
      io.OptionBuilder().setTransports(['polling']).setExtraHeaders(
          {'x-public-key': options.client.getPublicKey()}).build(),
    );

    _socket!.onConnect((_) {
      debugPrint('Socket.io Conectado.');
      _socket!.emit('process:connect', {'processId': processId});
    });

    _socket!.on('process:message', (data) {
      final response = _SocketResponse.fromJson(data);

      if (!response.success) {
        return options.onError?.call(ClientError.fromError(response.toJson()));
      }

      if (response.event == SocketEvents.partialPayment) {
        return options.onPartialPayment?.call(ClientResponse(
          code: ClientCode.paymentPartial,
          message: response.message,
          result: PartialPaymentResponseData.fromJson(response.result),
        ));
      }

      if (response.event == SocketEvents.fundsReceived) {
        return options.onSuccess?.call(ClientResponse(
          code: ClientCode.paymentSuccess,
          message: response.message,
          result: QrResponseData.fromJson(response.result),
        ));
      }
    });

    _socket!.onConnectError((error) {
      options.onError?.call(ClientError.fromError(
        error,
        code: ClientCode.socketConnectionError,
        message: I18n.t.errors.socketConnectionError,
      ));
      disconnect();
    });

    _socket!.onDisconnect((reason) {
      debugPrint('Socket.io Desconectado: $reason');
      _socket = null;
    });
  }

  void disconnect() {
    if (_socket != null) {
      _socket!.disconnect();
      _socket = null;
    }
  }
}
